src/proxy/dns.rs

```diff
@@ -385,30 +392,86 @@ impl DnsStore {
 
         addrs
     }
+
+    async fn forward(&self, client: &Workload, request: &Request) -> Result<Answer, LookupError> {
+        // Increment counter for all requests.
+        self.metrics.increment(&DnsRequest {
+            request,
+            source: Some(client),
+            destination: None,
+        });
+
+        // Increment counter for forwarded requests.
+        self.metrics.increment(&ForwardedRequest {
+            request,
+            source: Some(client),
+        });
+
+        // Record the forwarded request duration when the function exits.
+        let start = std::time::Instant::now();
+        let _forwarded_duration = self.metrics.defer(|metrics| {
+            metrics.record(
+                &ForwardedDuration {
+                    request,
+                    source: Some(client),
+                },
+                start.elapsed(),
+            );
+        });
+
+        match self.forwarder.forward(client, request).await {
+            Ok(answer) => Ok(answer),
+            Err(e) => {
+                // Increment counter for forwarding failures.
+                self.metrics.increment(&ForwardedFailure {
+                    request,
+                    source: Some(client),
+                });
+
+                Err(e)
+            }
+        }
+    }
 }
 
 #[async_trait::async_trait]
 impl Resolver for DnsStore {
     async fn lookup(&self, request: &Request) -> Result<Answer, LookupError> {
         // Find the client workload.
         let client = match self.find_client(to_canonical(request.src())) {
-            None => return Err(LookupError::ResponseCode(ResponseCode::ServFail)),
+            None => {
+                // Increment request counter.
+                self.metrics.increment(&DnsRequest {
+                    request,
+                    source: None,
+                    destination: None,
+                });
+
+                return Err(LookupError::ResponseCode(ResponseCode::ServFail));
+            }
             Some(client) => client,
         };
 
         // Make sure the request is for IP records. Anything else, we forward.
         let record_type = request.query().query_type();
         if !is_record_type_supported(record_type) {
-            return self.forwarder.forward(&client, request).await;
+            return self.forward(&client, request).await;
         }
 
         // Find the service for the requested host.
         let requested_name = Name::from(request.query().name().clone());
         let Some(service_match) = self.find_server(&client, &requested_name) else {
             // Unknown host. Forward to the upstream resolver.
-            return self.forwarder.forward(&client, request).await;
+            return self.forward(&client, request).await;
         };
 
+        // Increment counter for all requests.
+        self.metrics.increment(&DnsRequest {
```

Comment:
Looks like it's being incremented with similar labels either in `forward` or here.
Wondering if it should be placed once at line 455?

Reply:
Ah good catch. This call site should have the destination set. Updated.